### PR TITLE
Modify qml.grad so that it stores and makes accessible the value of the intermediate forward pass

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

If we would like to track the cost function during optimization, the ability to access the value of the cost function computed during the optimization step is ideal, rather than requiring the cost function to be recomputed manually and externally.

However, none of the optimizers explicitly compute the forward pass/cost function internally. So there is no explicit cost value that can be provided.

This is a quirk of Autograd - unlike PyTorch and TensorFlow, which enforce (via their UI) the need to manually perform a forward pass before you can access the backward pass, Autograd has no such restriction. Instead, calling ``autograd.grad()`` causes Autograd to perform both the forward pass and the backward pass at once.

**Description of the Change:**

* Modifies `grad_fn = qml.grad(cost_fn)` so that the implicit forward pass value is stored as an attribute, `grad_fn.forward`. This allows for both:

  - backcompatibility with the existing UI

  - the ability to easily extract the intermediate forward pass value if needed.

  For example, consider the following script (which remains compatible with the `master` branch):

  ```python
  from pennylane import numpy as np
  import pennylane as qml

  def cost_fn(x, y):
      return 2*np.sin(x[0])*np.exp(-x[1]) + x[0]**3 + np.cos(y)

  params = np.array([0.1, 0.5], requires_grad=True)
  data = np.array(0.65, requires_grad=False)
  grad_fn = qml.grad(cost_fn)
  ```

  Before the `grad_fn` is evaluated, the stored forward pass is not available:
  ```pycon
  >>> grad_fn.forward
  None
  ```

  Evaluating the gradient makes it accessible:
  ```pycon
  >>> print("Gradient:", grad_fn(params, data))
  Gradient: (array([ 1.23700107, -0.12110406]),)
  >>> print("Stored forward pass value:", grad_fn.forward)
  Stored forward pass value: 0.9181878546693896
  ```

  Double checking this agrees with cost function evaluation:
  ```pycon
  >>> print("Cost function:", cost_fn(params, data))
  Cost function: 0.9181878546693896
  ```

**Benefits:**

* Significant saving in number of quantum evaluations required if needing to track the cost function value during backpropagation.

* As a side effect of this change, the new `qml.grad` implementation is slightly leaner - there is no more function wrapping happening if the user does not provide `argnums`, which leads to a slightly cleaner and more performant UI.

* Note that this *only* applies to the autograd interface - PyTorch and TF both require manual forward passes before the backward pass is accessible.

**Possible Drawbacks:**

* We replicate the logic of `autograd.grad` here, rather than importing it directly. Normally I would not recommend modifying an external package function, since we would have to maintain it, and might miss breaking changes/bug fixes. However, in this case it should be okay, because:

  - The logic we are replicating is only two lines

  - Autograd is not maintained, so we do not expect any changes or bugfixes

* This PR does not yet allow the cost function to be returned from the optimizers, but it lays the necessary groundwork. A follow up PR should be added to modify the optimizers, to (optionally) accept a callback function `callback_fn(params, cost, grad)`, allowing the user to arbitrarily access and process the parameters, cost value, and gradient value at each step.

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by trbromley** [comment]
[ch1086]

File: pennylane/__init__.py [hunk 1]
```diff
@@ -227,13 +230,21 @@ def circuit():
     raise DeviceError("Device does not exist. Make sure the required plugin is installed.")
 
 
-def grad(func, argnum=None):
+make_vjp = unary_to_nary(_make_vjp)
```
Comment (trbromley) [inline pennylane/__init__.py:233]: What is `unary_to_nary`? :laughing:

File: pennylane/__init__.py [hunk 2]
```diff
@@ -227,13 +230,21 @@ def circuit():
...
     the ``argnum`` keyword argument is passed.
 
+    When the output gradient function is executed, both the forward pass
+    *and* the backward pass will be performed in order to
+    compute the gradient. The value of the forward pass is available via the
+    :attr:`~.forward` property.
+
     Args:
         func (function): a plain QNode, or a Python function that contains
             a combination of quantum and classical nodes
```
Comment (trbromley) [inline pennylane/__init__.py:250]: Do we still need to refer to `argnum` as a keyword argument below? `qml.grad(cost_fn, 0)` seems to work.

File: pennylane/__init__.py [hunk 3]
```diff
@@ -250,31 +261,77 @@ def grad(func, argnum=None):
...
-    if argnum is not None:
-        # for backwards compatibility with existing code
-        # that manually specifies argnum
-        return _grad(func, argnum)
 
-    def _gradient_function(*args, **kwargs):
-        """Inspect the arguments for differentiability, and
-        compute the autograd gradient function with required argnums
-        dynamically.
+    def __init__(self, fun, argnum=None, *args, **kwargs):
```
Comment (trbromley) [inline pennylane/__init__.py:265]: I'm curious why we now introduce `*args` and `**kwargs`?

File: pennylane/__init__.py [hunk 4]
```diff
@@ -250,31 +261,77 @@ def grad(func, argnum=None):
...
-        in differentiability.
+        self._fun = fun
+        self._argnum = argnum
+        self._grad_args = args
+        self._grad_kwargs = kwargs
+
+        if self._argnum is not None:
+            # If the differentiable argnum is provided, we can construct
+            # the gradient function at once during initialization
+            self._grad_fn = grad._grad_with_forward(fun, argnum=argnum, *args, **kwargs)
```
Comment (trbromley) [inline pennylane/__init__.py:277]: Looking at its signature, can `_grad_with_forward` support additional args and kwargs?

Comment (trbromley) [inline pennylane/__init__.py:277]: More out of curiosity since it is a static method, but why `grad._grad_with_forward` and not `self._grad_with_forward`?

File: tests/test_classical_gradients.py [hunk 5]
```diff
@@ -231,6 +231,60 @@ def test_linear(self, tol):
...
+
+        # change the parameters
+        params2 = np.array([1.4, 1.0, 2.0], requires_grad=True)
+        grad = grad_fn(params2)
+
+        res = grad_fn.forward
+        expected = cost(params2)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+
```
Comment (trbromley) [inline tests/test_classical_gradients.py:287]: I couldn't see anywhere in the tests a check for `qml.grad` where we have multiple variables, some with `requires_grad=True` and some with `requires_grad=False`, might be worth adding? (I could see this for `qml.jacobian()` though).

**Review by trbromley — COMMENTED** [review]
Thanks @josh146 it looks great! Just had a few questions.

File: pennylane/__init__.py (hunk 1 above)
Comment (josh146) [reply pennylane/__init__.py:233]: I don't know 😆 For some reason, you have to do this before you can use this function.

File: pennylane/__init__.py (hunk 2 above)
Comment (josh146) [reply pennylane/__init__.py:250]: Yeah, I'm not sure why this was done. I can move it.

File: pennylane/__init__.py (hunk 3 above)
Comment (josh146) [reply pennylane/__init__.py:265]: You're right, this can be removed 🙂 I think I was trying to be extra cautious when coding this up

File: pennylane/__init__.py (hunk 4 above)
Comment (josh146) [reply pennylane/__init__.py:277]: Good catch! I will remove this

Comment (josh146) [reply pennylane/__init__.py:277]: Ah, this is just convention. There isn't really much difference between the two different ways of calling a static method.
```suggestion
            self._grad_fn = self._grad_with_forward(fun, argnum=argnum, *args, **kwargs)
```

File: tests/test_classical_gradients.py (hunk 5 above)
Comment (josh146) [reply tests/test_classical_gradients.py:311]: Have added another test!

File: pennylane/__init__.py [hunk 6]
```diff
@@ -19,7 +19,10 @@
 import pkg_resources
 
 import numpy as _np
-from autograd import grad as _grad
+from autograd.wrap_util import unary_to_nary
+from autograd.core import make_vjp as _make_vjp, make_jvp as _make_jvp
+from autograd.extend import vspace
+
```
Comment (thisac) [inline pennylane/__init__.py:25]:
```suggestion
```

File: pennylane/__init__.py [hunk 7]
```diff
@@ -250,31 +259,76 @@ def grad(func, argnum=None):
...
+        if not vspace(ans).size == 1:
+            raise TypeError(
+                "Grad only applies to real scalar-output functions. "
+                "Try jacobian, elementwise_grad or holomorphic_grad."
+            )
 
-    return _gradient_function
+        grad_value = vjp(vspace(ans).ones())
+        return grad_value, ans
```
Comment (thisac) [inline pennylane/__init__.py:331]: Nice! So these are basically the main lines that are changed...

File: tests/test_classical_gradients.py [hunk 8]
```diff
@@ -231,6 +231,84 @@ def test_linear(self, tol):
...
+        def cost(x):
+            return np.sum(np.sin(x) * x[0] ** 3)
+
+        grad_fn = qml.grad(cost)
+        params = np.array([0.5, 1.0, 2.0], requires_grad=True)
+        res = grad_fn(params)
+        expected = autograd.grad(cost)(params)
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
```
Comment (thisac) [inline tests/test_classical_gradients.py:260]: Nice with `rtol=0`. :fire:

File: tests/test_classical_gradients.py [hunk 9]
```diff
@@ -231,6 +231,84 @@ def test_linear(self, tol):
...
+        expected = cost(params)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # change the parameters
+        params2 = np.array([1.4, 1.0, 2.0], requires_grad=True)
+        grad = grad_fn(params2)
+
+        res = grad_fn.forward
+        expected = cost(params2)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
```
Comment (thisac) [inline tests/test_classical_gradients.py:285]: :+1:

**Review by thisac — APPROVED** [review]
I think this looks great @josh146. Really nice addition and implementation. :100:

**Review by trbromley — APPROVED** [review]
Awesome @josh146, thanks!
